tests/Integration/Behaviour/Features/Context/Domain/Product/DuplicateProductFeatureContext.php

```diff
@@ -48,6 +53,36 @@ public function duplicate(string $productReference, string $newProductReference)
         $this->getSharedStorage()->set($newProductReference, $newProductId->getValue());
     }
 
+    /**
+     * @When I bulk duplicate following products:
+     *
+     * @param TableNode $productsList
+     */
+    public function bulkDuplicate(TableNode $productsList): void
+    {
+        $productIds = [];
+        foreach ($productsList->getColumnsHash() as $productInfo) {
+            $productIds[] = $this->getSharedStorage()->get($productInfo['reference']);
+        }
+
+        try {
+            $newProductIds = $this->getCommandBus()->handle(new BulkDuplicateProductCommand($productIds));
+        } catch (ProductException $e) {
+            $this->setLastException($e);
+            return;
+        }
+
+        foreach($newProductIds as $newProductId) {
+            foreach ($productsList->getColumnsHash() as $productInfo) {
+                $oldProduct = $this->getProductForEditing($productInfo['reference']);
+                $newProduct = $this->getQueryBus()->handle(new GetProductForEditing($newProductId->getValue()));
+                if ($oldProduct->getDetails()->getReference() === $newProduct->getDetails()->getReference()) {
```

Comment:
I'm not sure associating the products by their reference field is that accurate since multiple products might share the same reference Although I don't know what we should use to detect the matching...

Except returning the association in the handler, via an associative array or an object I'm not sure